File: app/src/main/java/com/krystianwsul/checkme/gui/projects/ProjectListFragment.java
package com.krystianwsul.checkme.gui.projects;


import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.Loader;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.gui.AbstractFragment;
import com.krystianwsul.checkme.gui.FabUser;
import com.krystianwsul.checkme.gui.SelectionCallback;
import com.krystianwsul.checkme.loaders.ProjectListLoader;
import com.krystianwsul.checkme.persistencemodel.SaveService;
import com.krystianwsul.treeadapter.ModelNode;
import com.krystianwsul.treeadapter.TreeModelAdapter;
import com.krystianwsul.treeadapter.TreeNode;
import com.krystianwsul.treeadapter.TreeNodeCollection;
import com.krystianwsul.treeadapter.TreeViewAdapter;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.TreeMap;

public class ProjectListFragment extends AbstractFragment implements LoaderManager.LoaderCallbacks<ProjectListLoader.Data>, FabUser {
    private static final String SELECTED_PROJECT_IDS = "selectedProjectIds";

    private ProgressBar mProjectListProgress;
    private TextView mEmptyText;
    private RecyclerView mProjectListRecycler;

    @Nullable
    private FloatingActionButton mProjectListFab;

    @Nullable
    private TreeViewAdapter mTreeViewAdapter;

    private Integer mDataId;

    private final SelectionCallback mSelectionCallback = new SelectionCallback() {
        @Override
        protected void unselect() {
            Assert.assertTrue(mTreeViewAdapter != null);

            mTreeViewAdapter.unselect();
        }

        @Override
        protected void onMenuClick(MenuItem menuItem) {
            Assert.assertTrue(mTreeViewAdapter != null);

            List<TreeNode> selected = mTreeViewAdapter.getSelectedNodes();
            Assert.assertTrue(!selected.isEmpty());

            List<ProjectListAdapter.ProjectNode> projectNodes = Stream.of(selected)
                    .map(treeNode -> ((ProjectListAdapter.ProjectNode) treeNode.getModelNode()))
                    .collect(Collectors.toList());

            Set<String> projectIds = Stream.of(projectNodes)
                    .map(projectNode -> projectNode.mProjectData.getId())
                    .collect(Collectors.toSet());

            switch (menuItem.getItemId()) {
                case R.id.action_project_delete:
                    Assert.assertTrue(mDataId != null);

                    for (TreeNode treeNode : selected) {
                        Assert.assertTrue(treeNode != null);

                        ProjectListAdapter.ProjectNode projectNode = (ProjectListAdapter.ProjectNode) treeNode.getModelNode();

                        projectNode.remove();

                        decrementSelected();
                    }

                    DomainFactory.getDomainFactory().setProjectEndTimeStamps(getActivity(), mDataId, SaveService.Source.GUI, projectIds);

                    break;
                default:
                    throw new UnsupportedOperationException();
            }
        }

        @Override
        protected void onFirstAdded() {
            Assert.assertTrue(mTreeViewAdapter != null);

            ((AppCompatActivity) getActivity()).startSupportActionMode(this);

            mTreeViewAdapter.onCreateActionMode();

            mActionMode.getMenuInflater().inflate(R.menu.menu_projects, mActionMode.getMenu());

            updateFabVisibility();
        }

        @Override
        protected void onSecondAdded() {

        }

        @Override
        protected void onOtherAdded() {

        }

        @Override
        protected void onLastRemoved() {
            Assert.assertTrue(mTreeViewAdapter != null);

            mTreeViewAdapter.onDestroyActionMode();

            updateFabVisibility();
        }

        @Override
        protected void onSecondToLastRemoved() {

        }

        @Override
        protected void onOtherRemoved() {

        }
    };

    @NonNull
    private Set<String> mSelectedProjectIds = new HashSet<>();

    @NonNull
    public static ProjectListFragment newInstance() {
        return new ProjectListFragment();
    }

    public ProjectListFragment() {

    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (savedInstanceState != null && savedInstanceState.containsKey(SELECTED_PROJECT_IDS)) {
            List<String> selectedProjectIds = savedInstanceState.getStringArrayList(SELECTED_PROJECT_IDS);
            Assert.assertTrue(selectedProjectIds != null);

            mSelectedProjectIds = new HashSet<>(selectedProjectIds);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_project_list, container, false);

        mProjectListProgress = view.findViewById(R.id.projectListProgress);
        Assert.assertTrue(mProjectListProgress != null);

        mEmptyText = view.findViewById(R.id.empty_text);
        Assert.assertTrue(mEmptyText != null);

        mProjectListRecycler = view.findViewById(R.id.projectListRecycler);
        Assert.assertTrue(mProjectListRecycler != null);

        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        mProjectListRecycler.setLayoutManager(new LinearLayoutManager(getActivity()));

        getLoaderManager().initLoader(0, null, this);
    }

    @Override
    public Loader<ProjectListLoader.Data> onCreateLoader(int id, Bundle args) {
        return new ProjectListLoader(getActivity());
    }

    @Override
    public void onLoadFinished(Loader<ProjectListLoader.Data> loader, ProjectListLoader.Data data) {
        Assert.assertTrue(data != null);

        mDataId = data.DataId;

        mProjectListProgress.setVisibility(View.GONE);
        if (data.getProjectDatas().isEmpty()) {
            mProjectListRecycler.setVisibility(View.GONE);
            mEmptyText.setVisibility(View.VISIBLE);
            mEmptyText.setText(R.string.projects_empty);
        } else {
            mProjectListRecycler.setVisibility(View.VISIBLE);
            mEmptyText.setVisibility(View.GONE);
        }

        if (mTreeViewAdapter != null)
            mSelectedProjectIds = Stream.of(mTreeViewAdapter.getSelectedNodes())
                    .map(treeNode -> ((ProjectListAdapter.ProjectNode) treeNode.getModelNode()).mProjectData.getId())
                    .collect(Collectors.toSet());

        mTreeViewAdapter = new ProjectListAdapter(getActivity()).initialize(data.getProjectDatas());
        mProjectListRecycler.setAdapter(mTreeViewAdapter);

        mSelectionCallback.setSelected(mTreeViewAdapter.getSelectedNodes().size());

        updateFabVisibility();
    }

    @Override
    public void onLoaderReset(Loader<ProjectListLoader.Data> loader) {

    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        if (mTreeViewAdapter != null)
            mSelectedProjectIds = Stream.of(mTreeViewAdapter.getSelectedNodes())
                    .map(treeNode -> ((ProjectListAdapter.ProjectNode) treeNode.getModelNode()).mProjectData.getId())
                    .collect(Collectors.toSet());

        outState.putStringArrayList(SELECTED_PROJECT_IDS, new ArrayList<>(mSelectedProjectIds));
    }

    @Override
    public void setFab(@NonNull FloatingActionButton floatingActionButton) {
        mProjectListFab = floatingActionButton;

        mProjectListFab.setOnClickListener(v -> startActivity(ShowProjectActivity.newIntent(getActivity())));

        updateFabVisibility();
    }

    private void updateFabVisibility() {
        if (mProjectListFab == null)
            return;

        if (mDataId != null && !mSelectionCallback.hasActionMode()) {
            mProjectListFab.show();
        } else {
            mProjectListFab.hide();
        }
    }

    @Override
    public void clearFab() {
        if (mProjectListFab == null)
            return;

        mProjectListFab.setOnClickListener(null);

        mProjectListFab = null;
    }

    private class ProjectListAdapter implements TreeModelAdapter {
        @NonNull
        private final Context mContext;

        private List<ProjectNode> mProjectNodes;

        private TreeViewAdapter mTreeViewAdapter;
        private TreeNodeCollection mTreeNodeCollection;

        ProjectListAdapter(@NonNull Context context) {
            mContext = context;
        }

        @NonNull
        private TreeViewAdapter initialize(@NonNull TreeMap<String, ProjectListLoader.ProjectData> projectDatas) {
            mProjectNodes = Stream.of(projectDatas.values())
                    .map(projectData -> new ProjectNode(this, projectData))
                    .collect(Collectors.toList());

            mTreeViewAdapter = new TreeViewAdapter(this);
            mTreeNodeCollection = new TreeNodeCollection(mTreeViewAdapter);
            mTreeViewAdapter.setTreeNodeCollection(mTreeNodeCollection);

            mTreeNodeCollection.setNodes(Stream.of(mProjectNodes)
                    .map(projectNode -> projectNode.initialize(mTreeNodeCollection))
                    .collect(Collectors.toList()));

            return mTreeViewAdapter;
        }

        @NonNull
        @Override
        public Holder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            LayoutInflater layoutInflater = LayoutInflater.from(mContext);

            View rowProject = layoutInflater.inflate(R.layout.row_project, parent, false);
            Assert.assertTrue(rowProject != null);

            return new Holder(rowProject);
        }

        @Override
        public boolean hasActionMode() {
            return mSelectionCallback.hasActionMode();
        }

        @Override
        public void incrementSelected() {
            mSelectionCallback.incrementSelected();
        }

        @Override
        public void decrementSelected() {
            mSelectionCallback.decrementSelected();
        }

        private void remove(@NonNull ProjectNode projectNode) {
            Assert.assertTrue(mProjectNodes.contains(projectNode));

            mProjectNodes.remove(projectNode);

            TreeNode treeNode = projectNode.mTreeNode;

            mTreeNodeCollection.remove(treeNode);
        }

        private class ProjectNode implements ModelNode {
            @NonNull
            private final ProjectListAdapter mProjectListAdapter;

            @NonNull
            private final ProjectListLoader.ProjectData mProjectData;

            private TreeNode mTreeNode;

            ProjectNode(@NonNull ProjectListAdapter projectListAdapter, @NonNull ProjectListLoader.ProjectData projectData) {
                mProjectListAdapter = projectListAdapter;
                mProjectData = projectData;
            }

            @NonNull
            TreeNode initialize(@NonNull TreeNodeCollection treeNodeCollection) {
                mTreeNode = new TreeNode(this, treeNodeCollection, false, mSelectedProjectIds.contains(mProjectData.getId()));
                mTreeNode.setChildTreeNodes(new ArrayList<>());
                return mTreeNode;
            }

            @Override
            public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder) {
                Holder holder = (Holder) viewHolder;

                holder.mProjectName.setText(mProjectData.getName());

                holder.mProjectUsers.setText(mProjectData.getUsers());

                holder.itemView.setOnClickListener(mTreeNode.getOnClickListener());

                holder.itemView.setOnLongClickListener(mTreeNode.getOnLongClickListener());

                if (mTreeNode.isSelected())
                    holder.itemView.setBackgroundColor(ContextCompat.getColor(getActivity(), R.color.selected));
                else
                    holder.itemView.setBackgroundColor(Color.TRANSPARENT);
            }

            @Override
            public int getItemViewType() {
                return 0;
            }

            @Override
            public boolean selectable() {
                return true;
            }

            @Override
            public void onClick() {
                mContext.startActivity(ShowProjectActivity.newIntent(mContext, mProjectData.getId()));
            }

            @Override
            public boolean visibleWhenEmpty() {
                return true;
            }

            @Override
            public boolean visibleDuringActionMode() {
                return true;
            }

            @Override
            public boolean separatorVisibleWhenNotExpanded() {
                return false;
            }

            @Override
            public int compareTo(@NonNull ModelNode modelNode) {
                Assert.assertTrue(modelNode instanceof ProjectNode);

                ProjectNode projectNode = (ProjectNode) modelNode;

                return mProjectData.getId().compareTo(projectNode.mProjectData.getId());
            }

            void remove() {
                mProjectListAdapter.remove(this);
            }
        }
    }

    private static class Holder extends RecyclerView.ViewHolder {
        @NonNull
        final TextView mProjectName;

        @NonNull
        final TextView mProjectUsers;

        Holder(View itemView) {
            super(itemView);

            mProjectName = itemView.findViewById(R.id.project_name);
            Assert.assertTrue(mProjectName != null);

            mProjectUsers = itemView.findViewById(R.id.project_users);
            Assert.assertTrue(mProjectUsers != null);
        }
    }
}

File: app/src/main/java/com/krystianwsul/checkme/gui/projects/ProjectListFragment.kt
package com.krystianwsul.checkme.gui.projects


import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.CustomItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.jakewharton.rxrelay3.BehaviorRelay
import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.databinding.FragmentProjectListBinding
import com.krystianwsul.checkme.domainmodel.extensions.clearProjectEndTimeStamps
import com.krystianwsul.checkme.domainmodel.extensions.setProjectEndTimeStamps
import com.krystianwsul.checkme.domainmodel.update.AndroidDomainUpdater
import com.krystianwsul.checkme.gui.base.AbstractFragment
import com.krystianwsul.checkme.gui.base.ActionModeListener
import com.krystianwsul.checkme.gui.base.SnackbarListener
import com.krystianwsul.checkme.gui.dialogs.RemoveInstancesDialogFragment
import com.krystianwsul.checkme.gui.instances.ShowTaskInstancesActivity
import com.krystianwsul.checkme.gui.main.FabUser
import com.krystianwsul.checkme.gui.main.MainActivity
import com.krystianwsul.checkme.gui.tasks.ShowTasksActivity
import com.krystianwsul.checkme.gui.tree.AbstractHolder
import com.krystianwsul.checkme.gui.tree.AbstractModelNode
import com.krystianwsul.checkme.gui.tree.BaseAdapter
import com.krystianwsul.checkme.gui.tree.HolderType
import com.krystianwsul.checkme.gui.tree.delegates.multiline.MultiLineDelegate
import com.krystianwsul.checkme.gui.tree.delegates.multiline.MultiLineModelNode
import com.krystianwsul.checkme.gui.tree.delegates.multiline.MultiLineRow
import com.krystianwsul.checkme.gui.utils.ResettableProperty
import com.krystianwsul.checkme.gui.utils.SelectionCallback
import com.krystianwsul.checkme.gui.widgets.MyBottomBar
import com.krystianwsul.checkme.utils.animateVisibility
import com.krystianwsul.checkme.viewmodels.ProjectListViewModel
import com.krystianwsul.checkme.viewmodels.getViewModel
import com.krystianwsul.common.utils.ProjectKey
import com.krystianwsul.treeadapter.*
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.kotlin.addTo
import io.reactivex.rxjava3.kotlin.plusAssign
import java.io.Serializable
import java.util.*

class ProjectListFragment : AbstractFragment(), FabUser {

    companion object {

        private const val SELECTED_PROJECT_IDS = "selectedProjectIds"

        private const val TAG_REMOVE_INSTANCES = "removeInstances"

        fun newInstance() = ProjectListFragment()
    }

    private var projectListFab: FloatingActionButton? = null

    lateinit var treeViewAdapter: TreeViewAdapter<AbstractHolder>
        private set

    private var data: ProjectListViewModel.Data? = null

    private val mainActivity get() = activity as MainActivity

    private val listener get() = activity as ProjectListListener

    private val selectionCallback = object : SelectionCallback() {

        override val activity get() = requireActivity()

        override fun getTreeViewAdapter() = treeViewAdapter

        override fun unselect(placeholder: TreeViewAdapter.Placeholder) = treeViewAdapter.unselect(placeholder)

        override val bottomBarData by lazy {
            Triple(listener.getBottomBar(), R.menu.menu_projects, listener::initBottomBar)
        }

        override fun onMenuClick(itemId: Int, placeholder: TreeViewAdapter.Placeholder): Boolean {
            checkNotNull(data)

            val projectIds = treeViewAdapter.selectedNodes
                    .map { (it.modelNode as ProjectListAdapter.ProjectNode).projectData.id }
                    .toHashSet()

            check(projectIds.isNotEmpty())

            val projectKey by lazy { projectIds.single() }

            when (itemId) {
                R.id.projectsMenuShowInstances -> startActivity(ShowTaskInstancesActivity.getIntent(
                        ShowTaskInstancesActivity.Parameters.Project(projectKey),
                ))
                R.id.projectsMenuShowTasks ->
                    startActivity(ShowTasksActivity.newIntent(ShowTasksActivity.Parameters.Project(projectKey)))
                R.id.projectsMenuEdit ->
                    startActivity(ShowProjectActivity.newIntent(requireContext(), projectKey))
                R.id.projectsMenuDelete -> RemoveInstancesDialogFragment.newInstance(projectIds)
                        .also { it.listener = deleteInstancesListener }
                        .show(childFragmentManager, TAG_REMOVE_INSTANCES)
                else -> throw UnsupportedOperationException()
            }

            return true
        }

        override fun onFirstAdded(placeholder: TreeViewAdapter.Placeholder, initial: Boolean) {
            (activity as AppCompatActivity).startSupportActionMode(this)

            updateFabVisibility()

            mainActivity.onCreateActionMode(actionMode!!)

            super.onFirstAdded(placeholder, initial)
        }

        override fun onLastRemoved(placeholder: TreeViewAdapter.Placeholder) {
            updateFabVisibility()

            mainActivity.onDestroyActionMode()
        }

        override fun getItemVisibilities(): List<Pair<Int, Boolean>> {
            val selectedNodes = getTreeViewAdapter().selectedNodes
            check(selectedNodes.isNotEmpty())

            val single = selectedNodes.size == 1

            return listOf(
                    R.id.projectsMenuShowInstances,
                    R.id.projectsMenuShowTasks,
                    R.id.projectsMenuEdit,
            ).map { it to single }
        }
    }

    private var selectedProjectIds: Set<ProjectKey.Shared> = setOf()

    private val isVisible = BehaviorRelay.createDefault(false)
    private lateinit var projectListViewModel: ProjectListViewModel

    private val deleteInstancesListener: (Serializable, Boolean) -> Unit = { projectIds, removeInstances ->
        checkNotNull(data)

        @Suppress("UNCHECKED_CAST")
        AndroidDomainUpdater.setProjectEndTimeStamps(
                projectListViewModel.dataId.toFirst(),
                projectIds as Set<ProjectKey.Shared>,
                removeInstances,
        )
                .observeOn(AndroidSchedulers.mainThread())
                .flatMapMaybe { mainActivity.showSnackbarRemovedMaybe(projectIds.size).map { _ -> it } }
                .flatMapCompletable {
                    AndroidDomainUpdater.clearProjectEndTimeStamps(projectListViewModel.dataId.toFirst(), it)
                }
                .subscribe()
                .addTo(createDisposable)
    }

    private val bindingProperty = ResettableProperty<FragmentProjectListBinding>()
    private var binding by bindingProperty

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (savedInstanceState?.containsKey(SELECTED_PROJECT_IDS) == true) {
            val selectedProjectIds = savedInstanceState.getParcelableArrayList<ProjectKey.Shared>(SELECTED_PROJECT_IDS)!!

            this.selectedProjectIds = HashSet(selectedProjectIds)
        }

        (childFragmentManager.findFragmentByTag(TAG_REMOVE_INSTANCES) as? RemoveInstancesDialogFragment)?.listener = deleteInstancesListener
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) =
            FragmentProjectListBinding.inflate(inflater, container, false).also { binding = it }.root

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.projectListRecycler.layoutManager = LinearLayoutManager(activity)

        projectListViewModel = getViewModel<ProjectListViewModel>().apply {
            viewCreatedDisposable += isVisible.subscribe { if (it) start() else stop() }

            viewCreatedDisposable += data.subscribe { onLoadFinished(it) }
        }
    }

    private fun onLoadFinished(data: ProjectListViewModel.Data) {
        this.data = data

        val hide = mutableListOf<View>(binding.projectListProgress)
        val show: View

        if (data.projectDatas.isEmpty()) {
            hide += binding.projectListRecycler
            show = binding.projectListEmptyTextInclude.emptyTextLayout

            binding.projectListEmptyTextInclude
                    .emptyText
                    .setText(R.string.projects_empty)
        } else {
            show = binding.projectListRecycler
            hide += binding.projectListEmptyTextInclude.emptyTextLayout
        }

        animateVisibility(listOf(show), hide, data.immediate)

        if (this::treeViewAdapter.isInitialized) {
            selectedProjectIds = treeViewAdapter.selectedNodes
                    .map { (it.modelNode as ProjectListAdapter.ProjectNode).projectData.id }
                    .toSet()

            treeViewAdapter.updateDisplayedNodes {
                (treeViewAdapter.treeModelAdapter as ProjectListAdapter).initialize(data.projectDatas)
                selectionCallback.setSelected(treeViewAdapter.selectedNodes.size, it, false)
            }
        } else {
            val projectListAdapter = ProjectListAdapter()
            projectListAdapter.initialize(data.projectDatas)
            treeViewAdapter = projectListAdapter.treeViewAdapter

            binding.projectListRecycler.apply {
                adapter = treeViewAdapter
                itemAnimator = CustomItemAnimator()
            }

            treeViewAdapter.updateDisplayedNodes {
                selectionCallback.setSelected(treeViewAdapter.selectedNodes.size, it, true)
            }
        }

        updateFabVisibility()
        updateSelectAll()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        if (this::treeViewAdapter.isInitialized)
            selectedProjectIds = treeViewAdapter.selectedNodes
                    .asSequence()
                    .map { (it.modelNode as ProjectListAdapter.ProjectNode).projectData.id }
                    .toSet()

        outState.putParcelableArrayList(SELECTED_PROJECT_IDS, ArrayList(selectedProjectIds))
    }

    override fun setFab(floatingActionButton: FloatingActionButton) {
        projectListFab = floatingActionButton

        projectListFab!!.setOnClickListener { startActivity(ShowProjectActivity.newIntent(requireActivity())) }

        updateFabVisibility()

        isVisible.accept(true)
    }

    private fun updateFabVisibility() {
        projectListFab?.run {
            if (data != null && !selectionCallback.hasActionMode) {
                show()
            } else {
                hide()
            }
        }
    }

    private fun updateSelectAll() = mainActivity.setProjectSelectAllVisibility(treeViewAdapter.displayedNodes.isNotEmpty())

    override fun clearFab() {
        isVisible.accept(false)

        projectListFab = null
    }

    override fun onDestroyView() {
        bindingProperty.reset()

        super.onDestroyView()
    }

    private inner class ProjectListAdapter : BaseAdapter(), ActionModeCallback by selectionCallback {

        val treeViewAdapter = TreeViewAdapter(
                this,
                TreeViewAdapter.PaddingData(R.layout.row_group_list_fab_padding, R.id.paddingProgress),
        )

        private lateinit var projectNodes: MutableList<ProjectNode>

        override lateinit var treeNodeCollection: TreeNodeCollection<AbstractHolder>
            private set

        fun initialize(projectDatas: SortedMap<ProjectKey.Shared, ProjectListViewModel.ProjectData>) {
            projectNodes = projectDatas.values
                    .map(::ProjectNode)
                    .toMutableList()

            treeNodeCollection = TreeNodeCollection(treeViewAdapter)
            treeViewAdapter.setTreeNodeCollection(treeNodeCollection)

            treeNodeCollection.nodes = projectNodes.map { it.initialize(treeNodeCollection) }
        }

        inner class ProjectNode(val projectData: ProjectListViewModel.ProjectData) :
                AbstractModelNode(),
                MultiLineModelNode {

            override val holderType = HolderType.MULTILINE

            override val id = projectData.id

            public override lateinit var treeNode: TreeNode<AbstractHolder>
                private set

            fun initialize(treeNodeCollection: TreeNodeCollection<AbstractHolder>): TreeNode<AbstractHolder> {
                treeNode = TreeNode(this, treeNodeCollection, selectedProjectIds.contains(projectData.id))
                treeNode.setChildTreeNodes(ArrayList())
                return treeNode
            }

            override val name get() = MultiLineRow.Visible(projectData.name)

            override val details get() = MultiLineRow.Visible(projectData.users, R.color.textSecondary)

            override val isSelectable = true

            override val parentNode: ModelNode<AbstractHolder>? = null

            override val delegates by lazy { listOf(MultiLineDelegate(this)) }

            override val widthKey
                get() = MultiLineDelegate.WidthKey(
                        0,
                        false,
                        false,
                        false
                )

            override fun onClick(holder: AbstractHolder) =
                    startActivity(ShowTasksActivity.newIntent(ShowTasksActivity.Parameters.Project(projectData.id)))

            override fun compareTo(other: ModelNode<AbstractHolder>): Int {
                check(other is ProjectNode)

                return projectData.id.compareTo(other.projectData.id)
            }
        }
    }

    interface ProjectListListener : SnackbarListener, ActionModeListener {

        fun setProjectSelectAllVisibility(selectAllVisible: Boolean)

        fun getBottomBar(): MyBottomBar

        fun initBottomBar()
    }
}
